File: 758-convert-binary-search-tree-to-sorted-doubly-linked-list/convert-binary-search-tree-to-sorted-doubly-linked-list.js
/**
 * // Definition for a _Node.
 * function _Node(val, left, right) {
 *      this.val = val;
 *      this.left = left;
 *      this.right = right;
 *  };
 */

/**
 * @param {_Node} root
 * @return {_Node}
 */
var treeToDoublyList = function(root) {
    if (!root) return null

    let first = null
    let last = null

    function inOrder(node) {
        if (!node) return

        inOrder(node.left)

        if (last) {
            last.right = node
            node.left = last
        } else {
            first = node
        }

        last = node

        inOrder(node.right)
    }

    inOrder(root)

    first.left = last
    last.right = first

    return first
};

File: 758-convert-binary-search-tree-to-sorted-doubly-linked-list/convert-binary-search-tree-to-sorted-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""

class Solution:
    def treeToDoublyList(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None

        self.first = None
        self.last = None

        def inorder(node):
            if not node:
                return
            
            inorder(node.left)

            if self.last:
                self.last.right = node
                node.left = self.last
            else:
                self.first = node

            self.last = node

            inorder(node.right)

        inorder(root)

        self.first.left = self.last
        self.last.right = self.first

        return self.first